core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala

```diff
@@ -0,0 +1,772 @@
+      snContext.createTable(s"$table4", "column", dataDF.schema,
+        (props + ("PARTITION_BY" -> "col4")))
+      tablesToDrop += table4
+
+      snContext.createTable(s"$rtable5", "row", dataDF.schema,
+        (props -- Seq("PARTITION_BY") + ("REPLICATE" -> "true")))
+      tablesToDrop += rtable5
+
+      snContext.createTable(s"$rtable6", "row", dataDF.schema,
+        (props -- Seq("PARTITION_BY") + ("REPLICATE" -> "true")))
+      tablesToDrop += rtable6
+
+      dataDF.write.format("column").mode(SaveMode.Append).options(props).saveAsTable(table1)
+      tablesToDrop += table1
+
+      dataDF.write.insertInto(table2)
+      dataDF.write.insertInto(table3)
+      dataDF.write.insertInto(table4)
+      dataDF.write.insertInto(rtable5)
+      dataDF.write.insertInto(rtable6)
+    }
+
+
+    createBaseTables()
+
+    val index1 = s"${table1}_Idx_One"
+    val index2 = s"${table1}_Idx_Two"
+    val index3 = s"${table1}_Idx_Three"
+
+    doPrint("Verify index create and drop for various index types")
+
+    snContext.sql(s"create index $index1 on $table1 " +
+        s" (COL2, COL3) Options (colocate_with  '$table2')")
+    indexesToDrop += index1
+
+    snContext.sql(s"create index $index2 on $table1 " +
+        s" (COL1, COL3) Options (colocate_with  '$table3')")
+    indexesToDrop += index2
+
+    snContext.sql(s"create index $index3 on $table1 " +
+        s" (COL4) Options (colocate_with  '$table4')")
+    indexesToDrop += index3
+
+    val executeQ = QueryExecutor(snContext, false, false)
+    /*
+    executeQ(s"select t1.col2, t2.col3 from $table1 t1, $table2 t2, $table4 t4 " +
+        s"where t1.col2 = t2.col2 and t1.col3 = t2.col3 " +
+        s"and t1.col4 = t4.col4 and t1.col1 = t4.col2 ")
+*/
+
+    executeQ(s"select t1.col2, t2.col3 from $table1 t1, $table2 t2 " +
+        s"where t1.col2 = t2.col2 and t1.col3 = t2.col3 ") {
+      validateIndex(Seq(index1), table2)(_)
+    }
+
+    executeQ(s"select t1.col2, t2.col3 from $table1 t1, $table2 t2, $table3 t3 " +
+        s"where t1.col2 = t2.col2 and t1.col3 = t2.col3 " +
+        s"and t1.col1 = t3.col1 and t1.col3 = t3.col3") {
+      // t1 -> t2, t1 -> t3
+      validateIndex(Seq(index1), table2, table3)(_)
+    }
+
+    executeQ(s"select t1.col2, t4.col3 from $table1 t1, $table4 t4 " +
+        s"where t1.col4 = t4.col4 and t1.col1 = t4.col2 ") {
+      // still results into shuffle.
+      validateIndex(Seq(index3), table4)(_)
+    }
+
+    executeQ(s"select t1.col2, t2.col3 from $table1 t1, $table2 t2, $table4 t4 " +
+        s"where t1.col2 = t2.col2 and t1.col3 = t2.col3 " +
+        s"and t1.col4 = t4.col4 ") {
+      // t1 -> t2, t1 -> t4
+      validateIndex(Seq(index1), table2, table4)(_)
+    }
+
+    executeQ(s"select t1.col2, t2.col3 from $table1 t1, $table4 t4, $table2 t2 " +
+        s"where t1.col2 = t2.col2 and t1.col3 = t2.col3 " +
+        s"and t1.col4 = t4.col4 ") {
+      // t1 -> t4, t1 -> t2
+      validateIndex(Seq(index1), table2, table4)(_)
+    }
```

Comment:
@ymahajan only one colocated join is possible for a given query. Picking up another global index instead of a table will still result in shuffle. So, I don't see any advantage of picking up additional index other than load balancing on the 3rd table scan. But as members are anyway limited and if indexes belong to different redundancy zone then probably this is worth it otherwise not.